### PR TITLE
Fixed NullReferenceException in OperationComplexityMiddleware (#5278)

### LINKED ISSUE #5278: System.NullReferenceException thrown from the OperationComplexityMiddleware class
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

We recently enabled complexity calculation for the queries using following configuration:

```
.ModifyRequestOptions(x => 
{   
    x.Complexity.Enable = true;  
    x.Complexity.ApplyDefaults = true;  
    x.Complexity.DefaultComplexity = 1;  
    x.Complexity.DefaultResolverComplexity = 5;  
    x.Complexity.MaximumAllowed = int.MaxValue;  
})  
```

Unfortunately after rolling it out to production NullReferenceException started to occur in the log and it seems to be coming from the inside of the library (HotChocolate.Execution.Pipeline.OperationComplexityMiddleware).

### Steps to reproduce

It might be hard to track it down as I am not able to reproduce the issue. Following is the sample query along with variables that comes from the request that caused mentioned exception:

```
query MyQuery($typeNameVar: String!) {
  __type(name: $typeNameVar) {
    name
    fields {
      name
      isDeprecated
      type {
        name
        ofType {
          name
          kind
          ofType {
            name
          }
        }
      }
    }
  }
}

variables:

{"typeNameVar": "DeletedCurrency"}
```

It seems that issue is not related to the query itself because when I ran it on my dev machine or against the system where error originally occurred I am unable to reproduce it.

### Relevant log output

```shell
System.NullReferenceException:
   at HotChocolate.Execution.Pipeline.OperationComplexityMiddleware+<InvokeAsync>d__7.MoveNext (HotChocolate.Execution, Version=12.7.0.0, Culture=neutral, PublicKeyToken=null)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at HotChocolate.Execution.Pipeline.OperationCacheMiddleware+<InvokeAsync>d__4.MoveNext (HotChocolate.Execution, Version=12.7.0.0, Culture=neutral, PublicKeyToken=null)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at HotChocolate.Execution.Pipeline.DocumentValidationMiddleware+<InvokeAsync>d__4.MoveNext (HotChocolate.Execution, Version=12.7.0.0, Culture=neutral, PublicKeyToken=null)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at HotChocolate.Execution.Pipeline.DocumentParserMiddleware+<InvokeAsync>d__4.MoveNext (HotChocolate.Execution, Version=12.7.0.0, Culture=neutral, PublicKeyToken=null)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at HotChocolate.Execution.Pipeline.DocumentCacheMiddleware+<InvokeAsync>d__5.MoveNext (HotChocolate.Execution, Version=12.7.0.0, Culture=neutral, PublicKeyToken=null)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at HotChocolate.Execution.Pipeline.TimeoutMiddleware+<InvokeAsync>d__3.MoveNext (HotChocolate.Execution, Version=12.7.0.0, Culture=neutral, PublicKeyToken=null)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at HotChocolate.Execution.Pipeline.ExceptionMiddleware+<InvokeAsync>d__3.MoveNext (HotChocolate.Execution, Version=12.7.0.0, Culture=neutral, PublicKeyToken=null)
```

### Additional Context?

It seems that in most cases where this exception occurred requests was send manually from the BananaCakePop using port forwarding to the AKS cluster. At this moment I registered only 2 requests (out of around 630k) sent by other micro services that failed due to described issue.

Maybe there will be something obvious about the implementation from which the exception was thrown.

Thank you in advance for your support.

### Product

Hot Chocolate

### Version

12.7.0

### PR DESCRIPTION
**Pull request by marcioscpinheiro**
- Fixes the operation complexity cache key
- Change the comparison to the SyntaxComparison

Closes #5278

### CODE REVIEW

**Comment by michaelstaib** [comment]
The fix introduces a new bug as the cache id is for the request but we need one for the operation.

**Comment by michaelstaib** [comment]
Actually, it was correct as it was and this introduces a bug.

File: src/HotChocolate/Core/src/Execution/Pipeline/OperationComplexityMiddleware.cs [hunk 1]
```diff
@@ -115,17 +116,14 @@ public async ValueTask InvokeAsync(IRequestContext context)
 
             foreach (var analyzer in analyzers)
             {
-                if (analyzer.OperationDefinitionNode == operationDefinition)
+                if (analyzer.OperationDefinitionNode.Equals(operationDefinition, SyntaxComparison.Syntax))
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Execution/Pipeline/OperationComplexityMiddleware.cs:119]: A reference check should be sufficient here as we are dealing with the same document. Why would you need to compare the syntax tree?

File: src/HotChocolate/Core/src/Execution/Pipeline/OperationComplexityMiddleware.cs [hunk 2]
```diff
@@ -115,17 +116,14 @@ public async ValueTask InvokeAsync(IRequestContext context)
...
                     operationAnalyzer = analyzer.Analyzer;
-                }
 
-                _cache.TryAddAnalyzer(
-                    requestContext.CreateCacheId(
-                        CreateOperationId(
-                            requestContext.DocumentId!,
-                            analyzer.OperationDefinitionNode.Name?.Value)),
-                    analyzer.Analyzer);
+                    _cache.TryAddAnalyzer(
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Execution/Pipeline/OperationComplexityMiddleware.cs:123]: This is incorrect, we create all analyzers here not just for the current operation. This should in the very least have a `break`. What is your reasoning here?

**Review by michaelstaib — CHANGES_REQUESTED** [review]
This PR introduces issues, see my comments.

**Comment by michaelstaib** [comment]
Also, can you add a test that shows how to create the NullRefException?

**Comment by michaelstaib** [comment]
I have a test of issue one .... in the invoke method the cache ID is wrong...

**Comment by michaelstaib** [comment]
I have explored the issue more and there is more to it ... I am closing this issue in favor of another PR I have open that will tackle the caching issue more broadly.

**Comment by michaelstaib** [comment]
#6100

**Comment by michaelstaib** [comment]
We will still give you credits for the fix, thats why I merged you initial attempt.

**Comment by marcioscpinheiro** [comment]
Thanks Michael, really appreciate your quick response.

**Pull request merged**